src/request.js

```diff
@@ -0,0 +1,237 @@
+  if (!validateHeaders(headers)) {
+    throw new RequestError('Headers are invalid');
+  }
+  headers['user-agent'] = 'cloud-mirror' + versionString;
+
+  let port;
+  try {
+    if (urlParts.port) {
+      port = parseInt(urlParts.port, 10);
+    }
+  } catch (err) {
+    throw new RequestError('Cannot parse port'); 
+  }
+
+  let method = opts.method || 'GET';
+  method = method.toUpperCase();
+
+  assert(http.METHODS.indexOf(method) !== -1);
+  
+  if (opts.stream) {
+    assert(typeof opts.contentType === 'string',
+        'when supplying a request body, you must specify content-type');
+    assert(!opts.data, 'if supplying a stream, you must not supply data');
+  }
+
+  if (opts.data) {
+    assert(typeof opts.contentType === 'string',
+        'when supplying a request body, you must specify content-type');
+    assert(!opts.stream, 'if supplying a stream, you must not supply data');
+  }
+
+  switch (urlParts.protocol) {
+    case 'https:':
+      return makeRequest({
+        httpLib: https,
+        hostname: urlParts.hostname,
+        port: port || 443,
+        path: urlParts.path || '/',
+        headers: headers,
+        method: method,
+        stream: opts.stream,
+        data: opts.data,
+        contentType: opts.contentType,
+        timeout: opts.timeout || 60000,
+      });
+      break;
+    case 'http:':
+      return makeRequest({
+        httpLib: http,
+        hostname: urlParts.hostname,
+        port: port || 80,
+        path: urlParts.path || '/',
+        headers: headers,
+        method: method,
+        stream: opts.stream,
+        data: opts.data,
+        contentType: opts.contentType,
+        timeout: opts.timeout || 60000,
+      });
+      break;
+    default:
+      throw new RequestError('Invalid protocol: ' + urlParts.protocol);
+      break;
+  }
+  
+}
+
+module.exports = {
+  request,
+  RequestError,
+};
+
+async function makeRequest(opts) {
+  assert(typeof opts === 'object');
+  assert(typeof opts.httpLib === 'object');
+  assert(typeof opts.hostname === 'string');
+  assert(typeof opts.port === 'number');
+  assert(typeof opts.path === 'string');
+  assert(typeof opts.method === 'string');
+  assert(typeof opts.headers === 'object');
+  assert(typeof opts.timeout === 'number');
+  if (opts.stream) {
+    assert(typeof opts.contentType === 'string');
+    assert(typeof opts.stream === 'object');
+    assert(typeof opts.stream.pipe === 'function');
+  }
+  if (opts.data) {
+    assert(typeof opts.contentType === 'string');
+    assert(typeof opts.data.length === 'number');
+  }
+
+  return new Promise(async (resolve, reject) => {
+    let request = opts.httpLib.request({
+      hostname: opts.hostname,
+      port: opts.port,
+      path: opts.path,
+      method: opts.method,
+    });
+
+    // We don't want connections hanging around forever, so we'll set a fatal
+    // condition that they timeout if they aren't responded to quickly enough
+    request.setTimeout(opts.timeout);
```

Comment:
Why not set this in `opts.httpLib.request({` as allowed by recent node.js versions.
See docs: https://nodejs.org/api/http.html#http_http_request_options_callback

That way the timeout will still kick in if we end up with TIME_WAIT issues... and never get a socket.